File: hash_table/palindrome_permutation.rb
# @param {String} s
# @return {Boolean}
def can_permute_palindrome(s)
    map, count = Hash.new { |h, k| h[k] = 0 }, 0

    s.each_char { |c| map[c] += 1 }
    map.values.each { |v| count += 1 if v % 2 == 1 }

    count <= 1
end


# 266. Palindrome Permutation
# https://leetcode.com/problems/palindrome-permutation/description/


# Key Insight
# Only 1 char in a palindrom can have an odd count

File: hash_table/palindrome_permutation.py
import collections


class Solution:
    def canPermutePalindrome(self, s: str) -> bool:
        counts = collections.Counter(s)
        odd_c_count = 0

        for _, count in counts.items():
            if count % 2 == 1: odd_c_count += 1
        return odd_c_count <= 1


class Solution:
    def canPermutePalindrome(self, s: str) -> bool:
        counts = collections.Counter(s)

        return len(filter(lambda x: x % 2 == 1, counts.values())) <= 1

# 266. Palindrome Permutation
# https://leetcode.com/problems/palindrome-permutation/description/
